Onboarding: Adds analytics

emilylaguna opened the pull request:
| 📘 Project: #500 | 🛫 Depends on: #541 |
|:---:|:---:|


This adds the analytics for the onboarding flow. This reuses whichever existing events we have for the similar views. 

Since the "onboarding" flow isn't just 1 singular thing, and can be entered from a variety of different places, I have added an OnboardingFlow struct that helps keep track of which flow we're in and appends that to the events. 

The flows that I have defined so far are:
- `initial_onboarding`: When the app first launches, and the user is asked to login/create account
- `plus_upsell`: When the user taps on a locked feature or upsell dialog and is brought to the plus landing view
- `plus_account_upgrade`: When the user taps on an upgrade button and is brought directly to the purchase modal
- `plus_account_upgrade_needs_login`: When the user taps on an upgrade button but is logged out and needs to login, They are presented with the login first, then the modal
- `logged_out`: When the user is logged out and enters the login flow, this flow is the same as the onboarding flow

Each of these flows have their own potential sub-flows such as pushing to the login from the landing or skipping steps. 

## To test
1. Enable the onboardingUpdates and tracksLogging FeatureFlags
2. Add `Settings.shouldShowInitialOnboardingFlow = true` to the AppDelegate `didFinishLaunchingWithOptions`

**Enable local IAP purchasing using StoreKit**
1. Open the Xcode Project
3. Hit <kbd>Command</kbd>+<kbd>Shift</kbd>+<kbd>,</kbd> to open the Scheme editor
4. Click the options tab
5. Select the `Pocket Casts Configuration.storekit` file in the StoreKit configuration dropdown


### Initial Onboarding
1. Launch the app
2. ✅ Verify you see the login view
4. ✅ Verify you see: `🔵 Tracked: setup_account_shown ["flow": "initial_onboarding"]`
5. Tap the Not Now button
6. ✅ `🔵 Tracked: setup_account_dismissed ["flow": "initial_onboarding"]`
7. Relaunch the app
8. Swipe down to dismiss the view
9. ✅ `🔵 Tracked: setup_account_dismissed ["flow": "initial_onboarding"]`
10. Relaunch the app
11. Tap on the Sign Up button
12. ✅ `🔵 Tracked: setup_account_button_tapped ["flow": "initial_onboarding", "button": "create_account"]`
13. Tap back
14. Tap on the Login Button
15. ✅ `🔵 Tracked: setup_account_button_tapped ["flow": "initial_onboarding", "button": "sign_in"]`
16. Login to an account without plus
17. After signing in a syncing
18. ✅ `🔵 Tracked: plus_promotion_shown ["flow": "initial_onboarding", "source": "login"]`
19. Tap the Unlock button
20. ✅ `🔵 Tracked: plus_promotion_upgrade_button_tapped ["flow": "initial_onboarding", "source": "login"]`
21. ✅ `🔵 Tracked: select_payment_frequency_shown ["flow": "initial_onboarding"]`
22. Dismiss the modal
23. ✅ `🔵 Tracked: select_payment_frequency_dismissed ["flow": "initial_onboarding"]`
24. Open it again
25. Tap on the subscribe button
26. ✅ `🔵 Tracked: select_payment_frequency_next_button_tapped ["product": "com.pocketcasts.plus.yearly", "flow": "initial_onboarding"]`
27. Subscribe
28. ✅ `🔵 Tracked: purchase_successful ["is_free_trial_available": false, "product": "com.pocketcasts.plus.yearly", "is_free_trial_eligible": BOOL]`
29. ✅ `🔵 Tracked: welcome_shown ["flow": "initial_onboarding", "display_type": "plus"]`
30. Tap the Import Podcasts option
31. ✅ `🔵 Tracked: welcome_import_tapped ["display_type": "plus", "flow": "initial_onboarding"]`
32. ✅ `🔵 Tracked: onboarding_import_shown ["flow": "initial_onboarding"]`
33. Tap on any app
34. ✅ `🔵 Tracked: onboarding_import_app_selected ["flow": "initial_onboarding", "app": "NAME OF APP YOU TAPPED"]`
35. Tap on the Open button
36. ✅ `🔵 Tracked: onboarding_import_open_app_tapped ["app": "NAME OF APP", "flow": "initial_onboarding"]`
37. Go back to the app
38. Go back to the welcome view
39. Tap the Discover button (find my next...)
40. ✅ `🔵 Tracked: welcome_discover_tapped ["flow": "initial_onboarding", "display_type": "plus"]`

### Plus Upsell
1. Sign into an account without plus
    - Or go to Profile > Cog > Developer > Set to No Plus
2. Go to the Podcasts tab
3. Tap on the Create Folder button
4. ✅ `🔵 Tracked: plus_promotion_shown ["flow": "plus_upsell", "source": "folders"]`
5. Proceed through the various parts of the flow as outlined above and ✅ verify the flow property is `plus_upsell`
6. Tap the various upsell buttons: Profile tab, Locked settings, etc
7. ✅ Verify the "source" appears correctly


### Plus Account Upgrade
1. Sign into an account without plus
2. Go to the Profile Tab
3. Tap on the Account Bubble
4. Tap Upgrade to Plus
5. ✅ `🔵 Tracked: select_payment_frequency_shown ["flow": "plus_account_upgrade", "source": "account_details"]`
6. Complete the purchase and verify flow is set to `plus_account_upgrade` and the `source` is `account_details`
7. Go to Profile Tab > Settings Cog > Plus item
8. Tap the Upgrade button
9. ✅ `🔵 Tracked: select_payment_frequency_shown ["flow": "plus_account_upgrade", "source": "plus_details"]`
10. Complete the purchase and verify flow is set to `plus_account_upgrade` and the `source` is `account_details`

### Plus Account Upgrade Needs Login
1. Sign out
2. Go to Profile Tab > Settings Cog > Plus item
3. Tap the Upgrade Button
4. ✅ Verify you see the login view
5. ✅ `🔵 Tracked: setup_account_shown ["flow": "plus_account_upgrade_needs_login"]`
6. Login to your account without plus
7. ✅ `🔵 Tracked: plus_promotion_shown ["flow": "plus_account_upgrade_needs_login", "source": "login"]`
8. ✅ Verify the purchase modal opens automatically
9. ✅ `🔵 Tracked: select_payment_frequency_shown ["flow": "plus_account_upgrade_needs_login"]`
11. Complete the rest of the steps and ✅ verify the flow is `plus_account_upgrade_needs_login`

### Logged Out
1. Launch the app
2. Dismiss the onboarding prompt
3. Go to the Profile tab
4. Tap the Set up account bubble
5. ✅ `🔵 Tracked: setup_account_shown ["flow": "logged_out"]`
6. Complete the rest of the steps and ✅ verify the flow is `logged_out`

## Checklist

- [ ] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [ ] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

emilylaguna committed:
Add onboarding upgrade view events

emilylaguna committed:
Add generic protocol for the onboarding models

emilylaguna committed:
Add onboarding view model to the onboarding hosting controller

emilylaguna committed:
Inform the view model of dismiss swipes

emilylaguna committed:
Remove common code from the plus hosting controller

emilylaguna committed:
Add plus landing events

emilylaguna committed:
Make an onboarding wrapper for the MDC wrapper

emilylaguna committed:
Add an enum to differentiate between dismiss types

emilylaguna committed:
Update onboarding upgrade shown

emilylaguna committed:
Add plus purchase modal events

emilylaguna committed:
Don't log the dismiss event if we've purchased

emilylaguna committed:
make the welcome view using the make method

emilylaguna committed:
Add welcome view events

emilylaguna committed:
Add analytics to the welcome view model

emilylaguna committed:
Add import view events

emilylaguna committed:
Add import view analytics

emilylaguna committed:
Create the import view using the make method

emilylaguna committed:
Remove the onboarding upgrade events

emilylaguna committed:
Adds a helper struct to represent the current onboarding flow

emilylaguna committed:
Replace existing flow creators to use the new onboarding flow helper

emilylaguna committed:
Update existing onboarding track methods to use the OnboardingFlow

emilylaguna committed:
Reset the flow when dismissed

emilylaguna committed:
Remove unused code

jgcaruso reviewed (commented):
I was able to test all of the flows up until the purchase. I got stuck in an apple id login loop, not sure if I just don't have some configuration set in xcode to do fake purchases?

But everything tested well up to that point. All of the `flow` properties were set correctly in the tracks event logging. Just a few things I noticed that might be useful to track:

1. The `purchase_cancelled` event doesn't include `flow` when cancelling purchase
2. `plus_account_upgrade` flow is used from both the Settings and Profile screens. Should they have separate flows for tracking or instead should `upgradeSource` be provided to specify where they came from?

I also had a weird behaviour that only happened once on a previous PR but I've been able to reproduce multiple times here (might not be a change here, but probably somewhere on the base branch?)

From settings -> Pocket Casts plus -> Upgrade -> login (enter login details and sign in) -> Plus sheet appears and then the purchase options sheet slides up immediately.

emilylaguna commented:
> I was able to test all of the flows up until the purchase. I got stuck in an apple id login loop, not sure if I just don't have some configuration set in xcode to do fake purchases?

I forgot to include the steps for this, I updated the testing steps to include setting up the StoreKit configuration file before running.


> The purchase_cancelled event doesn't include flow when cancelling purchase

That's to be expected, I only included the flow on all the new/reused events. It's not necessary to also add it to the purchase events since when we create a funnel we can determine which flow the purchase was started from using the previous events.


> I also had a weird behaviour that only happened once on a previous PR but I've been able to reproduce multiple times here (might not be a change here, but probably somewhere on the base branch?)

This is expected. In order to purchase any plan you need to be logged in first, but a few flows allow you to view the upsell view with an unlock/upgrade button before logging in. So in these instances you are brought to the login view, and after logging in/creating an account, since you've already tapped an upgrade button before, we "auto tap" it for you and bring you straight to the next step. This also occurs if you are logged out and tap on the create folder button, or one of the locked items in settings.

emilylaguna committed:
Add a method to reset the flow instead of modifying the property directly

emilylaguna committed:
Add the source to the Onboarding Flow tracking method

emilylaguna committed:
Add a way to update the flow analytics source outside of the initial settings

emilylaguna committed:
Set the onboarding source for the account and plus details prompt

emilylaguna committed:
Remove the upgradeSource code since it's handled in the onboarding flow now

emilylaguna committed:
Update the login source, and include the account_created

emilylaguna committed:
Add a source to the plus account prompt since it can be shown from a few places

emilylaguna committed:
Update the source for the account prompt

emilylaguna committed:
Reset the flow if the purchase modal is dismissed

emilylaguna commented:
> plus_account_upgrade flow is used from both the Settings and Profile screens. Should they have separate flows for tracking or instead should upgradeSource be provided to specify where they came from?

Good idea, I've added a source for those 2 areas. I've also refactored how the upgradeSource logic works, and moved it into the OnboardingFlow so it's tracks globally, and removed all the upgradeSource code since it's not needed anymore: 

https://github.com/Automattic/pocket-casts-ios/pull/542/files/588c8cea4fdfe49b4d7e25496a6d44e7670af0b2..57ea7ce586e9f4680dc841d7b78d5f3d2d35aff1

I've also updated the testing steps to include the new sources.

@jgcaruso another look please?

emilylaguna committed:
Fix a build error

jgcaruso reviewed (approved):
New changes continue to test as expected.

I was able to complete the purchase flows and test the Onboarding screen (import, discover, newsletter checkbox) and all were tracked as expected.
